File: src/main/java/com/webperformance/muse/measurements/StepDurationCollector.java
package com.webperformance.muse.measurements;

import org.musetest.core.*;
import org.musetest.core.events.*;
import org.musetest.core.step.*;

import java.util.*;

/**
 * Collects performance measurements on steps tagged for collection.
 *
 * @author Christopher L Merrill (see LICENSE.txt for license details)
 */
public class StepDurationCollector implements DataCollector, MuseEventListener
    {
    @Override
    public void initialize(MuseExecutionContext context)
        {
        context.addEventListener(this);
        }

    public List<Measurement> getMeasurements()
        {
        return _measurements;
        }

    @Override
    public void eventRaised(MuseEvent event)
        {
        if (event.getType().equals(MuseEventType.StartStep))
            {
            StepEvent start = (StepEvent) event;
            _start_time.put(start.getConfig(), start.getTimestampNanos());
            }
        else if (event.getType().equals(MuseEventType.EndStep))
            {
            StepEvent end = (StepEvent) event;
            Long start_time = _start_time.remove(end.getConfig());
            if (start_time != null)
                {
                final Measurement measurement = new Measurement();
                measurement.setValue(end.getTimestampNanos() - start_time);
                _measurements.add(measurement);
                }
            }
        }

    private HashMap<StepConfiguration, Long> _start_time = new HashMap<>();
    private List<Measurement> _measurements = new ArrayList<>();
    }

File: src/main/kotlin/com/webperformance/muse/measurements/StepDurationCollector.kt
package com.webperformance.muse.measurements

import mu.KotlinLogging
import org.musetest.core.MuseEvent
import org.musetest.core.MuseEventListener
import org.musetest.core.MuseEventType
import org.musetest.core.MuseExecutionContext
import org.musetest.core.context.initializers.ContextInitializerConfiguration
import org.musetest.core.datacollection.DataCollector
import org.musetest.core.events.StepEvent
import org.musetest.core.step.StepConfiguration
import java.util.*

private val logger = KotlinLogging.logger {}

/**
 * Collects performance measurements on all steps.
 *
 * @author Christopher L Merrill (see LICENSE.txt for license details)
 */
class StepDurationCollector : MuseEventListener, DataCollector
{
	private val startTime = HashMap<StepConfiguration, Long>()
	private val data = StepDurations()
	
	override fun configure(configuration: ContextInitializerConfiguration?)
	{
		// not expecting any configuration parameters yet
	}
	
	override fun getType(): String
	{
		return "com.webperformance.step-durations"
	}
	
	override fun initialize(context: MuseExecutionContext)
	{
		context.addEventListener(this)
	}
	
	override fun getData(): StepDurations
	{
		return data
	}
	
	override fun eventRaised(event: MuseEvent)
	{
		if (event.type == MuseEventType.StartStep)
		{
			val start = event as StepEvent
			startTime.put(start.config, start.timestampNanos)
		}
		else if (event.type == MuseEventType.EndStep)
		{
			val end = event as StepEvent
			val started: Long? = startTime.remove(end.config)
			if (started != null)
			{
				val measurement = Measurement(end.timestampNanos - started)
				data.durations.add(measurement)
			}
		}
		else if (event.type == MuseEventType.EndTest)
		{
			data.durations.iterator().forEach(operation = { measurement -> logger.error("measured: ${measurement.value}") })
		}
	}
}


